codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServerBuilderGenerator.kt

```diff
@@ -0,0 +1,487 @@
+import software.amazon.smithy.rust.codegen.server.smithy.ServerRuntimeType
+import software.amazon.smithy.rust.codegen.smithy.CodegenContext
+import software.amazon.smithy.rust.codegen.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.smithy.RustBoxTrait
+import software.amazon.smithy.rust.codegen.smithy.expectRustMetadata
+import software.amazon.smithy.rust.codegen.smithy.generators.StructureGenerator
+import software.amazon.smithy.rust.codegen.smithy.generators.builderSymbol
+import software.amazon.smithy.rust.codegen.smithy.isOptional
+import software.amazon.smithy.rust.codegen.smithy.isRustBoxed
+import software.amazon.smithy.rust.codegen.smithy.letIf
+import software.amazon.smithy.rust.codegen.smithy.makeOptional
+import software.amazon.smithy.rust.codegen.smithy.makeRustBoxed
+import software.amazon.smithy.rust.codegen.smithy.mapRustType
+import software.amazon.smithy.rust.codegen.smithy.rustType
+import software.amazon.smithy.rust.codegen.smithy.targetCanReachConstrainedShape
+import software.amazon.smithy.rust.codegen.smithy.traits.SyntheticInputTrait
+import software.amazon.smithy.rust.codegen.smithy.wrapMaybeConstrained
+import software.amazon.smithy.rust.codegen.util.hasTrait
+import software.amazon.smithy.rust.codegen.util.toPascalCase
+import software.amazon.smithy.rust.codegen.util.toSnakeCase
+
+// TODO Document differences:
+//     - This one takes in codegenContext.
+//     - Unlike in `BuilderGenerator.kt`, we don't add helper methods to add items to vectors and hash maps.
+//     - This builder is not `PartialEq`.
+//     - Always implements either From<Builder> for Structure or TryFrom<Builder> for Structure.
+class ServerBuilderGenerator(
+    private val codegenContext: CodegenContext,
+    private val shape: StructureShape,
+    private val takeInUnconstrainedTypes: Boolean = false,
+) {
+    private val model = codegenContext.model
+    private val symbolProvider = codegenContext.symbolProvider
+    private val constraintViolationSymbolProvider =
+        ConstraintViolationSymbolProvider(codegenContext.symbolProvider, model, codegenContext.serviceShape)
+    private val members: List<MemberShape> = shape.allMembers.values.toList()
+    private val structureSymbol = symbolProvider.toSymbol(shape)
+    private val moduleName = shape.builderSymbol(symbolProvider).namespace.split("::").last()
+    private val isBuilderFallible = StructureGenerator.serverHasFallibleBuilder(shape, model, symbolProvider, takeInUnconstrainedTypes)
+
+    fun render(writer: RustWriter) {
+        writer.docs("See #D.", structureSymbol)
+        writer.withModule(moduleName) {
+            renderBuilder(this)
+        }
+    }
+
+    private fun renderBuilder(writer: RustWriter) {
+        if (isBuilderFallible) {
+            Attribute.Derives(setOf(RuntimeType.Debug, RuntimeType.PartialEq)).render(writer)
+            // TODO(): `#[non_exhaustive] unless we commit to making builders of builders public.
```

Comment:
This can go away, you already added `#[non_exhaustive]` two lines below.

Reply:
The comment is supposed to convey that we should remove `#[non_exhaustive]` if we make builders of builders public.

Reply:
Oh ok.. It was not clear to me :)